### PR TITLE
Optimize usage under rustup.

### LINKED ISSUE #10986: Bypass rustup wrapper when invoking rustc
### Problem

I was just using flamegraph to profile a clean `cargo build` and observed that more than 4% of time was being spent parsing rustup configuration files. For a clean `cargo check` it was about 7%.

This seems to be because when you run `cargo`, it invokes rustup which then figures out which actual cargo binary to invoke (fine so far), then cargo invokes `rustc` many times and each time it does so, it's invoking rustup again, which has to go and reparse configuration files to figure out which rustc binary to invoke.

flamegraph

See `rustup_init::main` on the right (7.54%)

This was produced by running:

```sh
sudo apt install mold
cargo install flamegraph
git clone https://github.com/ebobby/simple-raytracer.git
cd simple-raytracer
cargo clean
flamegraph -- mold -run cargo check
```

To see the potential savings of bypassing rustup, we can set the `RUSTC` environment variable to point directly at rustc.

With rustup:
```
hyperfine --prepare 'cargo clean' 'cargo check'
Benchmark 1: cargo check
  Time (mean ± σ):      7.029 s ±  0.046 s    [User: 22.295 s, System: 2.782 s]
  Range (min … max):    6.939 s …  7.107 s    10 runs
```

Bypassing rustup:
```
RUSTC=`rustup which rustc` hyperfine --prepare 'cargo clean' 'cargo check'
Benchmark 1: cargo check
  Time (mean ± σ):      6.418 s ±  0.040 s    [User: 19.979 s, System: 2.409 s]
  Range (min … max):    6.347 s …  6.481 s    10 runs
```

So about a 9.5% speedup. I'd expect for crates with relatively few, large dependencies, the time spent by rustup would be less as a percentage. For crates with lots of small dependencies it could be more.

For a trivial (hello world) binary, a warm `cargo check` with a single line change is 101.5ms. If we set `RUSTC` to bypass rustup, this drops to 67.6ms.

The above times were all on my laptop. For an extra datapoint, I tried building nushell on a relatively powerful desktop with lots of RAM and CPUs. Cold `cargo check` went from 23.061 s ±  0.169 s to 22.313 s ±  0.188 s (3.4% speedup). A warm `cargo check` (with trivial one line change) went from 613.6 ms ±   3.6 ms to 582.3 ms ±   7.4 ms (5.4% speedup).

Steps performed by cargo to determine what `rustc` to run:
1. The value of the environment variable `RUSTC` if that's set
2. The value of `build.rustc` from `.cargo/config.toml` if set
3. *placeholder for extra step added by proposed solution and most alternatives*
4. Just run `rustc` and find it from `$PATH`

Other tools (e.g `rustdoc`) follow the same pattern.

`rustc` is actually a little different in that steps 1 and 2 are first performed for `rustc_wrapper` and `rustc_workspace_wrapper`. This doesn't affect the proposed solution or the alternatives.

## Proposed solution

Add step 3: Use tool (e.g. `rustc`) from the directory that contains the current `cargo` binary in preference to using `PATH`.

[Draft commit](https://github.com/davidlattimore/cargo/commit/655819db8bbce7d5dcae942b13211cc44794f27f) - still needs testing, updating of tests etc.

The main change to behaviour, would be a scenario like the following:

* A rust toolchain exists in `/aaa`
* A rust toolchain exists in `/bbb`
* `PATH=/aaa:/bbb` or `PATH=/aaa` - i.e. `/aaa` is the first or only toolchain on the path.
* User runs `/bbb/cargo build`
* Before, `/bbb/cargo` would have invoked `/aaa/rustc` because that's what's on the PATH
* After the proposed change, `/bbb/cargo` would invoke `/bbb/rustc`

While any change has the potential to break someone, the new behaviour actually seems less surprising to me. From the perspective of someone who only uses `cargo`, `rustc` could be considered an internal implementation detail and one might expect invoking `cargo` via an explicit path to use the associated `rustc` and other related tools.

### Alternative 1 - make rustup faster

This alone probably doesn't get us all the speedup we'd like. Even if we sped up rustup by a factor of 2, we'd still be spending a significant amount of time once we invoke it once for every `rustc` invocation.

### Alternative 2 - change rustup to set the `RUSTC` environment variable

This would break crates that set `build.rustc` in `.cargo/config.toml` since the value in `RUSTC` would override it. So this isn't really an option.

### Alternative 3a - rustup sets new environment variable that's then used by cargo

Rustup could set `DEFAULT_RUSTC`, which would be like `RUSTC` but would, if set, be used at step 3.

A downside is that this option treats rustc differently to the other tools that rustup wraps - unless we also add environment variables for other tools as well. e.g. `DEFAULT_RUSTDOC` - but that is messy and verbose.

### Alternative 3b - environment variable is the directory containing the tools

Similar to alternative 3a, but instead of having the new environment variable point to RUSTC, have it point to the directory. This would allow all tools to be treated consistently.

One downside of this option (and 3a) is that if the user bypasses the rustup wrapper by explicitly invoking cargo, then cargo would revert to invoking `rustup` for every call to `rustc`.

### Alternative 4 - cargo could run `rustup which rustc` once

If cargo ran `rustup which rustc`, it could then (if the command succeeded) use the result at step 3.

We'd still be running `rustup` twice - once to determine which `cargo` binary to invoke, then again within `cargo`. It's a lot better than invoking it N times though.

### Notes

_No response_

### PR DESCRIPTION
**Pull request by ehuss**
Closes #10986

This optimizes cargo when running under rustup to circumvent the rustup proxies. The rustup proxies introduce overhead that can make a noticeable difference.

The solution here is to identify if cargo would normally run `rustc` from PATH, and the current `rustc` in PATH points to something that looks like a rustup proxy (by comparing it to the `rustup` binary which is a hard-link to the proxy). If it detects this situation, then it looks for a binary in `$RUSTUP_HOME/toolchains/$TOOLCHAIN/bin/$TOOL`. If it finds the direct toolchain executable, then it uses that instead.

## Considerations

There have been some past attempts in the past to address this, but it has been a tricky problem to solve. This change has some risk because cargo is attempting to guess what the user and rustup wants, and it may guess wrong. Here are some considerations and risks for this:

* Setting `RUSTC` (as in https://github.com/rust-lang/rustup/pull/2958) isn't an option. This makes the `RUSTC` setting "sticky" through invocations of different toolchains, such as a cargo subcommand or build script which does something like `cargo +nightly build`.

* Changing `PATH` isn't an option, due to issues like https://github.com/rust-lang/rustup/issues/3036 where cargo subcommands would be unable to execute proxies (so things like `+toolchain` shorthands don't work).

* Setting other environment variables in rustup (as in https://github.com/rust-lang/rustup/pull/3207 which adds `RUSTUP_TOOLCHAIN_DIR` the path to the toolchain dir) comes with various complications, as there is risk that the environment variables could get out of sync with one another (like with `RUSTUP_TOOLCHAIN`), causing tools to break or become confused.

  There was some consideration in that PR for adding protections by using an encoded environment variable that could be cross-checked, but I have concerns about the complexity of the solution.

  We may want to go with this solution in the long run, but I would like to try a short term solution in this PR first to see how it turns out.

* This won't work for a `rustup-toolchain.toml` override with a [`path`](https://rust-lang.github.io/rustup/overrides.html#path) setting. Cargo will use the slow path in that case. In theory it could try to detect this situation, which may be an exercise for the future.

* Some build-scripts, proc-macros, or custom cargo subcommands may be doing unusual things that interfere with the assumptions made in this PR. For example, a custom subcommand could call a `cargo` executable that is not managed by rustup. Proc-macros may be executing cargo or rustc, assuming it will reach some particular toolchain. It can be difficult to predict what unusual ways cargo and rustc are being used. This PR (and its tests) tries to make extra sure that it is resilient even in unusual circumstances.

* The "dev" fallback in rustup can introduce some complications for some solutions to this problem. If a rustup toolchain does not have cargo, such as with a developer "toolchain link", then rustup will automatically call either the nightly, beta, or stable cargo if they are available. This PR should work correctly, since rustup sets the correct `RUSTUP_TOOLCHAIN` environment variable for the *actual* toolchain, not the one where cargo was executed from.

* Special care should be considered for dynamic linking. `LD_LIBRARY_PATH` (linux), `DYLD_LIBRARY_PATH` (macos), and `PATH` (windows) need to be carefully set so that `rustc` can find its shared libraries. Directly executing `rustc` has some risk that it will load the wrong shared libraries. There are some mitigations for this. macOS and Linux use rpath, and Windows looks in the same directory as `rustc.exe`. Also, rustup configures the dyld environment variables from the outer cargo. Finally, cargo also configures these (particularly for the deprecated compiler plugins).

* This shouldn't impact installations that don't use rustup.

* I've done a variety of testing on the big three platforms, but certainly nowhere exhaustive. 
    * One of many examples is making sure Clippy's development environment works correctly, which has special requirements for dynamic linking.

* There is risk about future rustup versions changing some assumptions made here. Some assumptions:
    * It assumes that if `RUSTUP_TOOLCHAIN` is set, then the proxy *always* runs exactly that toolchain and no other. If this changes, cargo could execute the wrong version. Currently `RUSTUP_TOOLCHAIN` is the highest priority [toolchain override](https://rust-lang.github.io/rustup/overrides.html) and is fundamental to how toolchain selection becomes "sticky", so I think it is unlikely to change.
    * It assumes rustup sets `RUSTUP_TOOLCHAIN` to a value that is exactly equal to the name of the toolchain in the `toolchains` directory. This works for user shorthands like `RUSTUP_TOOLCHAIN=nightly`, which gets converted to the full toolchain name. However, it does not work for `path` overrides (see above).
    * It assumes the `toolchains` directory layout is always `$RUSTUP_HOME/toolchains/$TOOLCHAIN`. If this changes, then I think the only consequence is that cargo will go back to the slow path.
    * It assumes downloading toolchains is not needed (since cargo running from the toolchain means it should already be downloaded).
    * It assumes there is no other environment setup needed (such as the dyld paths mentioned above).

  My hope is that if assumptions are no longer valid that the worst case is that cargo falls back to the slow path of running the proxy from PATH.

## Performance

This change won't affect the performance on Windows because rustup currently alters PATH to point to the toolchain directory. However, https://github.com/rust-lang/rustup/pull/3178 is attempting to remove that, so this PR will be required to avoid a performance penalty on Windows. That change is currently opt-in, and will likely take a long while to roll out since it won't be released until after the next release, and may be difficult to get sufficient testing.

I have done some rough performance testing on macOS, Windows, and Linux on a variety of different kinds of projects with different commands. The following attempts to summarize what I saw.

The timings are going to be heavily dependent on the system and the project. These are the values I get on my systems, but will likely be very different for everyone else.

The Windows tests were performed with a custom build of rustup with https://github.com/rust-lang/rustup/pull/3178 applied and enabled (stock rustup shows no change in performance as explained above).

The data is summarized in this spreadsheet: https://docs.google.com/spreadsheets/d/1zSvU1fQ0uSELxv3VqWmegGBhbLR-8_KUkyIzCIk21X0/edit?usp=sharing

`hello-world` has a particularly large impact of about 1.68 to 2.7x faster. However, a large portion of this overhead is related to running `rustc` at the start to discover its version and querying it for information. This is cached after the first run, so except for first-time builds, the effect isn't as noticeable. The "check with info" row is an benchmark that removes `target/debug/deps` but keeps the `.rustc_info.json` file.

Incremental builds are a bit more difficult to construct since it requires customizing the commands for each project. I only did an incremental test for cargo itself, running `touch src/cargo/lib.rs` and then `cargo check --lib`.

These measurements excluded the initial overhead of launching the rustup proxy to launch the initial cargo process. This was done just for simplicity, but it makes the test a little less characteristic of a typical usage, which will have some constant overhead for running the proxy.

These tests were done using [`hyperfine`](https://crates.io/crates/hyperfine) version 1.16.1. The macOS system was an M2 Max (12-thread). The Windows and Linux experiments were run on a AMD Ryzen Threadripper 2950X (32-thread). Rust 1.68.2 was used for testing. I can share the commands if people want to see them.

### CODE REVIEW

File: src/cargo/util/config/mod.rs [hunk 1]
```diff
@@ -1644,6 +1644,48 @@ impl Config {
...
+                // there is pretty good evidence that it will be a proxy.
+                let tool_resolved = paths::resolve_executable(Path::new(tool)).ok()?;
+                let rustup_resolved = paths::resolve_executable(Path::new("rustup")).ok()?;
+                let tool_meta = tool_resolved.metadata().ok()?;
+                let rustup_meta = rustup_resolved.metadata().ok()?;
+                // This works on the assumption that rustup and its proxies
+                // use hard links to a single binary. If rustup ever changes
+                // that setup, then I think the worst consequence is that this
+                // optimization will not work, and it will take the slow path.
+                if tool_meta.len() != rustup_meta.len() {
```
Comment (joshtriplett) [inline src/cargo/util/config/mod.rs:1672]: This seems hazardous. It's entirely possible for two binaries to have the same length without having the same contents. If you're checking for hardlinks, shouldn't this compare `ino` and `dev`, at least on Unix? (I'm not sure how to improve this on Windows, where I'm assuming these are copies rather than hardlinks, and even the modification time might not match.)

Comment (joshtriplett) [reply src/cargo/util/config/mod.rs:1672]: Following up: according to @ChrisDenton, you need to open both files, and then while both are open, call `GetFileInformationByHandleEx` and make sure both volume and ID are identical.

Comment (ChrisDenton) [reply src/cargo/util/config/mod.rs:1672]: Essentially you'd need to compare [`FILE_ID_INFO`](https://learn.microsoft.com/en-us/windows/win32/api/winbase/ns-winbase-file_id_info) structs. Though there are some nuances here and the point about keeping both file handles open is crucial because ids aren't otherwise *guaranteed* to be stable (see [[MS-FSCC]](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-fscc/98860416-1caf-4c80-a9ab-8d61e1ccf5a5) reference). See also [this LLVM bug](https://lists.llvm.org/pipermail/llvm-bugs/2014-December/037218.html).

File: src/cargo/util/config/mod.rs [hunk 2]
```diff
@@ -1644,6 +1644,48 @@ impl Config {
...
+                // calling the proxies.
+                //
+                // This is somewhat cautious trying to determine if it is safe
+                // to circumvent rustup, because there are some situations
+                // where users may do things like modify PATH, call cargo
+                // directly, use a custom rustup toolchain link without a
+                // cargo executable, etc. However, there is still some risk
+                // this may make the wrong decision in unusual circumstances.
+                //
+                // First, we must be running under rustup in the first place.
```
Comment (epage) [inline src/cargo/util/config/mod.rs:1660]: This feels a lot more complicated and brittle to me than rust-lang/cargo#10998.  In reviewing rust-lang/cargo#10998, the only downside I saw listed was that it wasn't being driven by rustup which this has the same problem.

Is there something I'm missing for why we'd prefer this route over rust-lang/cargo#10998?

File: src/cargo/util/config/mod.rs (hunk 1 above)
Comment (ehuss) [reply src/cargo/util/config/mod.rs:1672]: That's possible, but I'm concerned about the potential complexity or difficulty in getting that right. There are various situations where symlinks are used, and I can't guarantee that the files won't end up as a copy, or have issues across network mounts, for example. The file sizes are currently an order of magnitude different, and I think the chance for them being the same is very unlikely.

File: src/cargo/util/config/mod.rs (hunk 2 above)
Comment (ehuss) [reply src/cargo/util/config/mod.rs:1660]: Querying rustc is certainly a possibility, but the approach taken there incurs additional startup time for an initial cache. Adding a new flag to rustc has a fairly high bar, but adding a transparent optimization in cargo has no user-facing interaction so should be easier to move forward with. So, in terms of the global complexity (changes to cargo, rustc, and/or rustup, user-facing documentation, etc.), this seemed like the simplest solution with the least risk, and can receive benefits immediately rather than waiting a potentially very long time.

If this solution ends up having issues that one of the other solutions could address, then I think it would be worthwhile to re-investigate a different approach.

Comment (epage) [reply src/cargo/util/config/mod.rs:1660]: > Adding a new flag to rustc has a fairly high bar

Even if its just a new enumerated value for an existing flag?

> If this solution ends up having issues that one of the other solutions could address, then I think it would be worthwhile to re-investigate a different approach.

If I understand correctly, this solution could start failing and we'd never know it, right?

Comment (ehuss) [reply src/cargo/util/config/mod.rs:1660]: > Even if its just a new enumerated value for an existing flag?

Yea, new options are almost always added as an unstable option, and then it needs to go through the process of making a case for the compiler team to stabilize.

> If I understand correctly, this solution could start failing and we'd never know it, right?

It is possible, though I think unlikely in most cases. I think any major regressions would require a significant change in the design of rustup, and I think that is unlikely for the foreseeable future. I could add a test that requires rustup to be installed if that may help with that concern.

File: src/cargo/util/config/mod.rs [hunk 3]
```diff
@@ -1644,6 +1644,48 @@ impl Config {
...
+                let toolchain = self.get_env_os("RUSTUP_TOOLCHAIN")?;
+                // If the tool on PATH is the same as `rustup` on path, then
+                // there is pretty good evidence that it will be a proxy.
```
Comment (rbtcollins) [inline src/cargo/util/config/mod.rs:1663]: We have an exact list of the proxies we offer btw.

File: src/cargo/util/config/mod.rs (hunk 1 above)
Comment (rbtcollins) [reply src/cargo/util/config/mod.rs:1672]: I'm not sure if this matters but the reason the fs footprint of rustup looks quite different in different places is because of android (no hardlink support), bew (everything is a symlink to a symlink from our next release) (and I guess snap and other 3rd-party distributions can also differ to what one might expect from looking at our code).

In particular see https://github.com/rust-lang/rustup/pull/3137 for some context.

tl;dr: there is no guarantee that the proxy and rustup itself are the same file, even though that is our default installation logic.

On android they are symlinks.
On MacOSX with brew they are symlinks from our next release.

And the consequence of the tool not being a proxy is that someone has deliberately placed e.g. a 'rustc' wrapper that does something, which cargo would then not run.

Checking for the same length will detect every common situation where they are different except for two cases I can see: two different binaries, alike in length, and two different symlinks, alike in length.

For binaries I agree - its very unlikely that two different binaries the same length as rustup is large enough that the law of small numbers doesn't really apply.

For symlinks, I suggest doing a readlink on the file. It is cheap enough to still be a lot faster than rustup manifest parsing (which I plan to do something about someday, but its not top of the list, and even after, not running code we don't need to run is how we make things fast).

**Comment by rbtcollins** [comment]
I'm in favour of this as discussed.

One note - I'm looking to sharply constrain toolchain path based toolchains in the next release - removing the relative-path support entirely (or at least feature-flagging it to drive some feedback), and in a later release I hope to remove it. There's no tracking bug yet but some discord discussion.

All of which to say, I don't think you should worry about path based toolchains, other than rejecting toolchains that contain '/' or '\'.

File: src/cargo/util/config/mod.rs [hunk 4]
```diff
@@ -1644,6 +1644,48 @@ impl Config {
...
+                // optimization will not work, and it will take the slow path.
+                if tool_meta.len() != rustup_meta.len() {
+                    return None;
+                }
+                // Try to find the tool in rustup's toolchain directory.
+                let tool_exe = Path::new(tool).with_extension(env::consts::EXE_EXTENSION);
+                let toolchain_exe = home::rustup_home()
+                    .ok()?
+                    .join("toolchains")
+                    .join(&toolchain)
```
Comment (rbtcollins) [inline src/cargo/util/config/mod.rs:1680]: This could look at pretty random places on the filesystem with a path based toolchain. I suggest breaking out of this logic early based on `toolchain == 'none' || toolchain.contains('/') || toolchain.contains('\\')`

File: src/cargo/util/config/mod.rs (hunk 2 above)
Comment (rbtcollins) [reply src/cargo/util/config/mod.rs:1660]: I can add in that rustup has been consulted and we quite like this approach.

Comment (joshtriplett) [reply src/cargo/util/config/mod.rs:1660]: I definitely prefer a solution like this that doesn't add any additional invocations of rustc.

File: src/cargo/util/config/mod.rs (hunk 1 above)
Comment (ehuss) [reply src/cargo/util/config/mod.rs:1682]: > For symlinks, I suggest doing a readlink on the file. It is cheap enough to still be a lot faster than rustup manifest parsing (which I plan to do something about someday, but its not top of the list, and even after, not running code we don't need to run is how we make things fast).

Can you say more about why this would be needed? If the proxy symlink points at rustup, shouldn't they have the same size?

File: src/cargo/util/config/mod.rs (hunk 3 above)
Comment (ehuss) [reply src/cargo/util/config/mod.rs:1673]: Cargo is currently hard-coded to only use this for `rustc` and `rustdoc`. I added an assert to validate that requirement, and I think we can extend it in the future if needed. I don't think we quite yet need to have an exhaustive list for all the proxies (just to keep things simple for now).

File: src/cargo/util/config/mod.rs (hunk 4 above)
Comment (ehuss) [reply src/cargo/util/config/mod.rs:1690]: Good catch! I have added that.

File: src/cargo/util/config/mod.rs (hunk 1 above)
Comment (ehuss) [reply src/cargo/util/config/mod.rs:1682]: Maybe I'm still unclear, but this uses the standard `metadata` function which reads the target of a symlink (recursively). If `rustup` and `rustc` are symlinks to different things (with the same length path), they'll still have different length targets.

If that doesn't resolve your concern, can you show a specific example?  For example:

/usr/bin/rustc -> /usr/bin/rust-compiler
/usr/bin/rustup -> /usr/bin/rustup-thingy
/usr/bin/rust-compiler 669176
/usr/bin/rustup-thingy 8027337

These have the same length paths, but different length targets, so they should be treated as being different.

Comment (rbtcollins) [reply src/cargo/util/config/mod.rs:1682]: ok, so you're using fs::metadata(path).len(), not symlink_metadata ? Then I'm fine with that as-is

**Comment by rbtcollins** [comment]
Another complication but probably ok:

Users can specify `RUSTUP_TOOLCHAIN=nightly` : this is part of the Rustup UI. When a proxy is run with that set, the toolchain is resolved to e.g. `nightly-x86_64-pc-windows-msvc` and then the child process (e.g. cargo) will receive `RUSTUP_TOOLCHAIN=nightly-x86_64-pc-windows-msvc` which will be an actual directory in ~/.rustup/toolchains. 

I think this is ok because if someone runs cargo without the proxy, with that variable set, the directory won't exist, and the fallback path of just invoking the rustc proxy will invoke a rustup proxy, which will then perform resolution, ending up with the explicitly requested toolchain being run. 

If the user is running a cargo from a different toolchain, directly, with a RUSTUP_TOOLCHAIN variable set, then I think they can keep both pieces.

File: src/cargo/util/config/mod.rs [hunk 5]
```diff
@@ -1643,7 +1643,59 @@ impl Config {
...
+                // This works on the assumption that rustup and its proxies
+                // use hard links to a single binary. If rustup ever changes
+                // that setup, then I think the worst consequence is that this
+                // optimization will not work, and it will take the slow path.
+                if tool_meta.len() != rustup_meta.len() {
+                    return None;
+                }
+                // Try to find the tool in rustup's toolchain directory.
+                let tool_exe = Path::new(tool).with_extension(env::consts::EXE_EXTENSION);
+                let toolchain_exe = home::rustup_home()
```
Comment (weihanglo) [inline src/cargo/util/config/mod.rs:1687]: `home` crate still access `RUSTUP_HOME` via `std::env`. Will there be an inconsistency when people set their `RUSTUP_HOME` in `[env]`?

https://github.com/rust-lang/cargo/blob/39684ff0aa88fe856460bc8665fd6f649b065e94/crates/home/src/env.rs#L93

Comment (ehuss) [reply src/cargo/util/config/mod.rs:1687]: It shouldn't under normal circumstances, since the rustup proxies set `RUSTUP_HOME`, the `[env]` value will be ignored.

File: src/cargo/util/config/mod.rs [hunk 6]
```diff
@@ -1643,7 +1643,59 @@ impl Config {
     /// Looks for a path for `tool` in an environment variable or config path, defaulting to `tool`
     /// as a path.
     fn get_tool(&self, tool: &str, from_config: &Option<ConfigRelativePath>) -> PathBuf {
+        // This function is designed to only work with rustup proxies. This
+        // assert is to ensure that if it is ever used for something else in
+        // the future that you must ensure that it is a proxy-able tool, or if
+        // not then you need to use `maybe_get_tool` instead.
+        assert!(matches!(tool, "rustc" | "rustdoc"));
```
[the remark this answers is not in the document]
Comment (ehuss) [reply src/cargo/util/config/mod.rs]: Sure, I went ahead and added an enum.

**Review by weihanglo — APPROVED** [review]
This looks pretty good and simpler than other alternatives to me. Given that rustup folks agree on this change, Feel free to `r=weihanglo` if there is no further discussion needed with them.

**Comment by ehuss** [comment]
I think this should be ready to go. My intent is to get this out on nightly and hopefully get enough real-world testing to determine if there is a problem, and we can either fix it or back it out. Unfortunately nightly doesn't get enough testing in some of the more unusual environments, but hopefully it will be enough to be moderately confident it should be ok.

@bors r=weihanglo

File: src/cargo/util/config/mod.rs (hunk 5 above)
Comment (rbtcollins) [reply src/cargo/util/config/mod.rs:1691]: I think it would be a good idea to prohibit this before stablisation @ehuss

Comment (ehuss) [reply src/cargo/util/config/mod.rs:1691]: Sure, I went ahead and posted #12101.
